### ISSUE TITLE
False positives using valid-jsdoc and never return type

### ISSUE THREAD

**Issue by reduckted** [buzinas/tslint-eslint-rules#234]
|Package|Version|
|-------|-------|
|`tslint`| 5.4.2|
|`tsc`| 2.3.4|
|`tslint-eslint-rules`| 4.1.1|

### tslint.json

```json
{
    "rules": {
        "valid-jsdoc": [ true, { "requireReturn": false } ]
    },
    "rulesDirectory": [
        "node_modules/tslint-eslint-rules/dist/rules"
    ]
}
```

### Code

```typescript
/**
 * A test function.
 * @param {string} input The input.
 */
function test(input: string | undefined): void {
    if (input === undefined) {
        return fail('boom!');
    }

    console.log(input.length);
}

/**
 * Throws an error.
 * @param {string} message The error message.
 */
function fail(message: string): never {
    throw new Error(message);
}
```

### Actual Behaviour

The error `ERROR: index.ts[1, 1]: missing JSDoc @returns for function` is produced on the `test` function.

### Expected Behaviour

No errors are produced.

### More information

If I have a `void` function that returns the "result" of a call to a function with a return type of [`never`](https://basarat.gitbooks.io/typescript/docs/types/never.html), the `valid-jsdoc` rule flags that I'm missing an `@returns` tag. In this scenario the `return` statement is simply used as a way to tell the compiler that the function will exit at that point, rather than a value actually being returned.

If I change `return fail()` to just `fail()`, the `valid-jsdoc` error goes away, but then the typescript compiler flags an error because `input` is possibly undefined later on in the method.

**Comment by reduckted** [comment]
If the function declares a return type of `void`, the typescript compiler will produce errors if you try to return anything other than `undefined`, `void` or `never`. So instead of looking for `return` statements as an indication that the function returns something, maybe we could check the declared return type of the function. If it doesn't have a declared type, then we could look for `return` statements.
